doc/zarafa-monitor.1
'\" t
.\"     Title: zarafa-monitor
.\"    Author: [see the "Author" section]
.\" Generator: DocBook XSL Stylesheets v1.76.1 <http://docbook.sf.net/>
.\"      Date: August 2011
.\"    Manual: Zarafa user reference
.\"    Source: Zarafa 7.1
.\"  Language: English
.\"
.TH "ZARAFA\-MONITOR" "1" "August 2011" "Zarafa 7.1" "Zarafa user reference"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
zarafa-monitor \- Start the Zarafa monitor\&.
.SH "SYNOPSIS"
.HP \w'\fBzarafa\-monitor\fR\ 'u
\fBzarafa\-monitor\fR [\fIOPTION\fR...] [\fISERVERNAME\fR]
.SH "DESCRIPTION"
.PP
The monitor checks user mailbox sizes\&. When the quota is reached the monitor sends a quota email\&. When started, the program will keep running\&.
.PP
The defaults can be overwritten in the config file\&.
.PP
The default settings of the monitor program are defined in
\fBzarafa-monitor.cfg\fR(5)\&.
.SH "OPTIONS"
.PP
Various options are possible\&. None of these options are required\&.
.PP
\fB\-\-config\fR, \fB\-c\fR \fImonitor\&.cfg\fR
.RS 4
Specify the location of the monitor\&.cfg file\&. If the monitor\&.cfg file is not specified, the monitor will try to load
/etc/zarafa/monitor\&.cfg\&. If that file is not present, it will continue using it\*(Aqs default settings\&.
.sp
Default:
\fI/etc/zarafa/monitor\&.cfg\fR
.RE
.PP
\fB\-\-host\fR, \fB\-h\fR \fIpath\fR
.RS 4
Connect to the Zarafa server through
\fIpath\fR, e\&.g\&.
file:///path/to/socket\&. Default:
file:///var/run/zarafa\&.
.RE
.PP
\fB\-\-foreground\fR, \fB\-F\fR
.RS 4
Run in the foreground\&. Normally the monitor will daemonize and run in the background\&.
.RE
.PP
\fB\-\-ignore\-unknown\-config\-options\fR
.RS 4
When unknown options exists in the configuration file the monitor service will report an error and shutdown\&. Pass this option to ignore unknown options so the monitor service will start\&.
.RE
.SH "SIGNALS"
.PP
The following signals can be sent to the zarafa\-monitor process:
.PP
\fBHUP\fR
.RS 4
When the HUP signal is received, some options from the configuration file are reloaded\&. The reloadable options are listed in the
\fBzarafa-monitor.cfg\fR(5)
manual page\&.
.sp
Also, when using
\fIlog_method = file\fR, the logfile will be closed and a new logfile will be opened\&. You can use this signal in your logrotate system\&.
.RE
.PP
\fBTERM\fR
.RS 4
To gracefully let the service exit, the normal TERM signal is used\&.
.RE
.SH "USAGE"
.PP
The monitor checks user mailbox sizes\&. The monitor sends a quota email when the warning quota limit is reached\&. When started, the program will keep running\&. You may also use the init\&.d scripts:
.PP
\fB/etc/init\&.d/zarafa\-monitor\fR
[\fIstart\fR|
\fIstop\fR|
\fIrestart\fR]
.SH "AUTHOR"
.PP
Written by Zarafa\&.
.SH "SEE ALSO"
.PP

\fBzarafa-monitor.cfg\fR(5)
\fBzarafa-server\fR(1)


debian/changelog
giraffe (7.1.5.final-1~1.gbpdcf72d) UNRELEASED; urgency=low

  ** SNAPSHOT build @dcf72d11495083f2275fa118afad08bba3903cfb **

  [ Guido Günther ]
  * [6dbdff8] Change project name to giraffe
  * [ce5202f] Change maintainer to pkg-giraffe
  * [b691d5e] Switch DH_VERBOSE to false
  * [11a4382] Don't try to remove non existant directory

  [ Matthias Geerdsen ]
  * [2a1aa8f] fix libboost dependencies for sid
  * [7dd25fa] zarafa-webaccess-mobile is no longer shipped
  * [a90de06] use autotools-dev debhelpers for config.{sub,guess} updating
  * [d09323c] fix installation of libzarafasync
  * [072134d] rebase patch to new upstream
  * [11d8d22] add patch to fix casting problem with gzfile (from Mandriva)
  * [0f96759] add patch to fix php compatibilty issue pval -> zval
  * [ee82ed2] fix libvime0 for version in sid, add zarafa-contacts package
  * [6a5a9fb] add zarafa-contacts package
  * [c6d60f3] add /etc/mapi to zarafa-client
  * [e73e75a] fix typo in package name
  * [05eaf43] add patch to remove hex2bin() declaration, it is available in php 5.4.0

  [ Robert Kehl ]
  * Renaming zarafa to giraffe
  * Renaming zarafa-common to giraffe-common
  * Renaming zarafa-server to giraffe-server
  * Renaming zarafa-dbg to giraffe-dbg
  * Renaming zarafa-dev to giraffe-dev
  * Renaming zarafa-client to giraffe-client
  * Renaming zarafa-contacts to giraffe-contacts
  * Renaming zarafa-libs to giraffe-libs
  * Renaming zarafa-libarchiver to giraffe-libarchiver
  * Renaming zarafa-utils to giraffe-utils
  * Renaming zarafa-monitor to giraffe-monitor
  * Renaming zarafa-spooler to giraffe-spooler
  * Renaming zarafa-dagent to giraffe-dagent
  * Renaming zarafa-gateway to giraffe-gateway
  * Renaming zarafa-ical to giraffe-ical
  * Renaming zarafa to giraffe
  * Renaming zarafa-archiver to giraffe-archiver
  * Renaming zarafa-webaccess to giraffe-webaccess
  * Renaming zarafa-search to giraffe-search
  * Clarifying package descriptions
  * fixing file names in conf files
  * fix rename

 -- Robert Kehl <robertkehl@robertkehl.de>  Sat, 29 Mar 2014 14:39:50 +0100

giraffe (7.0.3~beta1-1~1.gbpe4c14c) UNRELEASED; urgency=low

  ** SNAPSHOT build @e4c14c384782481ae8adde0c376927d45caca787 **

  * [e4c14c3] Add gbp.conf

 -- Guido Günther <agx@sigxcpu.org>  Thu, 17 Nov 2011 20:30:30 +0100

zarafa (7.0.1-28228) unstable; urgency=low

  * new release, build using libgoogle-perftools

 -- John van der Kamp <john@memleak.office.zarafa.com>  Wed, 20 Jul 2011 16:20:09 +0200

zarafa (7.0.0-27791.1) unstable; urgency=low

  * Non-maintainer upload.
  * Make 3.0 (quilt) source format

 -- John van der Kamp <john@memleak.office.zarafa.com>  Fri, 15 Jul 2011 16:36:34 +0200

zarafa (7.0.0-27791) unstable; urgency=low

  * First OpenSource release of Zarafa Collaboration Platform for Debian.

 -- John van der Kamp <j.vdkamp@zarafa.com>  Fri, 15 Jul 2011 15:28:10 +0200


ECtools/zarafa-monitor/zarafa-monitor.cpp
/*
 * Copyright 2005 - 2014  Zarafa B.V.
 * 
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3, 
 * as published by the Free Software Foundation with the following additional 
 * term according to sec. 7:
 *  
 * According to sec. 7 of the GNU Affero General Public License, version
 * 3, the terms of the AGPL are supplemented with the following terms:
 * 
 * "Zarafa" is a registered trademark of Zarafa B.V. The licensing of
 * the Program under the AGPL does not imply a trademark license.
 * Therefore any rights, title and interest in our trademarks remain
 * entirely with us.
 * 
 * However, if you propagate an unmodified version of the Program you are
 * allowed to use the term "Zarafa" to indicate that you distribute the
 * Program. Furthermore you may use our trademarks where it is necessary
 * to indicate the intended purpose of a product or service provided you
 * use it in accordance with honest practices in industrial or commercial
 * matters.  If you want to propagate modified versions of the Program
 * under the name "Zarafa" or "Zarafa Server", you may only do so if you
 * have a written permission by Zarafa B.V. (to acquire a permission
 * please contact Zarafa at trademark@zarafa.com).
 * 
 * The interactive user interface of the software displays an attribution
 * notice containing the term "Zarafa" and/or the logo of Zarafa.
 * Interactive user interfaces of unmodified and modified versions must
 * display Appropriate Legal Notices according to sec. 5 of the GNU
 * Affero General Public License, version 3, when you propagate
 * unmodified or modified versions of the Program. In accordance with
 * sec. 7 b) of the GNU Affero General Public License, version 3, these
 * Appropriate Legal Notices must retain the logo of Zarafa or display
 * the words "Initial Development by Zarafa" if the display of the logo
 * is not reasonably feasible for technical reasons. The use of the logo
 * of Zarafa in Legal Notices is allowed for unmodified and modified
 * versions of the software.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *  
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 */

// zarafa-monitor.cpp : Defines the entry point for the console application.
//

#include "platform.h"


#include <stdio.h>
#include <stdlib.h>
#include <iostream>
#include <signal.h>

#include <mapi.h>
#include <mapix.h>
#include <mapiutil.h>
#include <mapidefs.h>

#include <mapiguid.h>
#include <ctype.h>
#include "my_getopt.h"

#include "ECScheduler.h"
#include "ECMonitorDefs.h"
#include "ECQuotaMonitor.h"

#include "CommonUtil.h"
#include "UnixUtil.h"
#include "ecversion.h"
#include "charset/localeutil.h"

using namespace std;

void deleteThreadMonitor(LPECTHREADMONITOR	lpThreadMonitor, bool base = false) {

	if(lpThreadMonitor == NULL)
		return;

	if(lpThreadMonitor->lpConfig)
		delete lpThreadMonitor->lpConfig;

	if(lpThreadMonitor->lpLogger)
		lpThreadMonitor->lpLogger->Release();

	if(base)
		delete lpThreadMonitor;
}

LPECTHREADMONITOR	m_lpThreadMonitor;

pthread_mutex_t		m_hExitMutex;
pthread_cond_t		m_hExitSignal;
pthread_t			mainthread;

HRESULT running_service( char* szPath );

void sighandle(int sig) {
	// Win32 has unix semantics and therefore requires us to reset the signal handler.
	signal(SIGTERM , sighandle);
	signal(SIGINT  , sighandle);	// CTRL+C

	if (m_lpThreadMonitor && m_lpThreadMonitor->bShutdown == false)// do not log multimple shutdown messages
		m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Termination requested, shutting down.");
	
	m_lpThreadMonitor->bShutdown = true;

	pthread_cond_signal(&m_hExitSignal);
}

void sighup(int signr) {
	// In Win32, the signal is sent in a seperate, special signal thread. So this test is
	// not needed or required.
	if (pthread_equal(pthread_self(), mainthread)==0)
		return;
	if (m_lpThreadMonitor) {
		if (m_lpThreadMonitor->lpConfig) {
			if (!m_lpThreadMonitor->lpConfig->ReloadSettings() && m_lpThreadMonitor->lpLogger)
				m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Unable to reload configuration file, continuing with current settings.");
		}

		if (m_lpThreadMonitor->lpLogger) {
			if (m_lpThreadMonitor->lpConfig) {
				char *ll = m_lpThreadMonitor->lpConfig->GetSetting("log_level");
				int new_ll = ll ? atoi(ll) : 2;
				m_lpThreadMonitor->lpLogger->SetLoglevel(new_ll);
			}

			m_lpThreadMonitor->lpLogger->Reset();
			m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_WARNING, "Log connection was reset");
		}
	}
}

// SIGSEGV catcher
#include <execinfo.h>

void sigsegv(int signr)
{
	void *bt[64];
	int i, n;
	char **btsymbols;

	if(!m_lpThreadMonitor || !m_lpThreadMonitor->lpLogger)
		goto exit;

	switch (signr) {
	case SIGSEGV:
		m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Caught SIGSEGV (%d), traceback:", signr);
		break;
	case SIGBUS:
		m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Caught SIGBUS (%d), possible invalid mapped memory access, traceback:", signr);
		break;
	case SIGABRT:
		m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Caught SIGABRT (%d), out of memory or unhandled exception, traceback:", signr);
		break;
	};

	n = backtrace(bt, 64);

	btsymbols = backtrace_symbols(bt, n);

	for (i = 0; i < n; i++) {
		if (btsymbols)
			m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "%016p %s", bt[i], btsymbols[i]);
		else
			m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "%016p", bt[i]);
	}

	m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "When reporting this traceback, please include Linux distribution name, system architecture and Zarafa version.");

exit:
	kill(0, signr);
}


void print_help(char *name) {
	cout << "Usage:\n" << endl;
	cout << name << " [-F] [-h|--host <serverpath>] [-c|--config <configfile>]" << endl;
	cout << "  -F\t\tDo not run in the background" << endl;
	cout << "  -h path\tUse alternate connect path (e.g. file:///var/run/socket).\n\t\tDefault: file:///var/run/zarafa" << endl;
	cout << "  -c filename\tUse alternate config file (e.g. /etc/zarafa-monitor.cfg)\n\t\tDefault: /etc/zarafa/monitor.cfg" << endl;
	cout << endl;
	cout << "  --ignore-unknown-config-options\tStart even if the configuration file contains invalid config options" << endl;
	cout << endl;
}

int main(int argc, char *argv[]) {

	HRESULT hr = hrSuccess;
	const char *szConfig = ECConfig::GetDefaultPath("monitor.cfg");
	char *szPath = NULL;
	int c;
	int daemonize = 1;
	bool bIgnoreUnknownConfigOptions = false;


	// Default settings
	const configsetting_t lpDefaults[] = {
		{ "smtp_server","localhost" },
		{ "server_socket", CLIENT_ADMIN_SOCKET },
		{ "run_as_user", "" },
		{ "run_as_group", "" },
		{ "pid_file", "/var/run/zarafa-monitor.pid" },
		{ "running_path", "/" },
		{ "log_method","file" },
		{ "log_file","/var/log/zarafa/monitor.log" },
		{ "log_level","2", CONFIGSETTING_RELOADABLE },
		{ "log_timestamp","1" },
		{ "sslkey_file", "" },
		{ "sslkey_pass", "", CONFIGSETTING_EXACT },
		{ "quota_check_interval", "15" },
		{ "mailquota_resend_interval", "1", CONFIGSETTING_RELOADABLE },
		{ "userquota_warning_template", "/etc/zarafa/quotamail/userwarning.mail", CONFIGSETTING_RELOADABLE },
		{ "userquota_soft_template", "", CONFIGSETTING_UNUSED },
		{ "userquota_hard_template", "", CONFIGSETTING_UNUSED },
		{ "companyquota_warning_template", "/etc/zarafa/quotamail/companywarning.mail", CONFIGSETTING_RELOADABLE },
		{ "companyquota_soft_template", "/etc/zarafa/quotamail/companysoft.mail", CONFIGSETTING_RELOADABLE },
		{ "companyquota_hard_template", "/etc/zarafa/quotamail/companyhard.mail", CONFIGSETTING_RELOADABLE },
		{ "servers", "" },
		{ NULL, NULL },
	};

	enum {
		OPT_HELP,
		OPT_HOST,
		OPT_CONFIG,
		OPT_FOREGROUND,
		OPT_IGNORE_UNKNOWN_CONFIG_OPTIONS
	};
	struct option long_options[] = {
		{ "help", 0, NULL, OPT_HELP },
		{ "host", 1, NULL, OPT_HOST },
		{ "config", 1, NULL, OPT_CONFIG },
		{ "foreground", 1, NULL, OPT_FOREGROUND },
		{ "ignore-unknown-config-options", 0, NULL, OPT_IGNORE_UNKNOWN_CONFIG_OPTIONS },
		{ NULL, 0, NULL, 0 }
	};

	if (!forceUTF8Locale(true))
		goto exit;

	while(1) {
		c = my_getopt_long_permissive(argc, argv, "c:h:iuFV", long_options, NULL);
		
		if(c == -1)
			break;
			
		switch(c) {
		case OPT_CONFIG:
		case 'c':
			szConfig = my_optarg;
			break;
		case OPT_HOST:
		case 'h':
			szPath = my_optarg;
			break;
		case 'i': // Install service
		case 'u': // Uninstall service
			break;
		case OPT_FOREGROUND:
		case 'F':
			daemonize = 0;
			break;
		case OPT_IGNORE_UNKNOWN_CONFIG_OPTIONS:
			bIgnoreUnknownConfigOptions = true;
			break;
		case 'V':
			cout << "Product version:\t" <<  PROJECT_VERSION_MONITOR_STR << endl
				 << "File version:\t\t" << PROJECT_SVN_REV_STR << endl;
			return 1;
		case OPT_HELP:
		default:
			print_help(argv[0]);
			return 1;
		}
	}

	m_lpThreadMonitor = new ECTHREADMONITOR;

	m_lpThreadMonitor->lpConfig = ECConfig::Create(lpDefaults);
	if (!m_lpThreadMonitor->lpConfig->LoadSettings(szConfig) || !m_lpThreadMonitor->lpConfig->ParseParams(argc-my_optind, &argv[my_optind], NULL) || (!bIgnoreUnknownConfigOptions && m_lpThreadMonitor->lpConfig->HasErrors())) {
		m_lpThreadMonitor->lpLogger = new ECLogger_File(EC_LOGLEVEL_FATAL, 0, "-"); // create fatal logger without a timestamp to stderr
		LogConfigErrors(m_lpThreadMonitor->lpConfig, m_lpThreadMonitor->lpLogger);
		hr = E_FAIL;
		goto exit;
	}

	mainthread = pthread_self();

	// setup logging
	m_lpThreadMonitor->lpLogger = CreateLogger(m_lpThreadMonitor->lpConfig, argv[0], "Zarafa-Monitor");

	if ((bIgnoreUnknownConfigOptions && m_lpThreadMonitor->lpConfig->HasErrors()) || m_lpThreadMonitor->lpConfig->HasWarnings())
		LogConfigErrors(m_lpThreadMonitor->lpConfig, m_lpThreadMonitor->lpLogger);


	// set socket filename
	if (!szPath)
		szPath = m_lpThreadMonitor->lpConfig->GetSetting("server_socket");

	signal(SIGTERM, sighandle);
	signal(SIGINT, sighandle);
	signal(SIGHUP, sighup);

	// SIGSEGV backtrace support
	stack_t st;
	struct sigaction act;

	memset(&st, 0, sizeof(st));
	memset(&act, 0, sizeof(act));

	st.ss_sp = malloc(65536);
	st.ss_flags = 0;
	st.ss_size = 65536;

	act.sa_handler = sigsegv;
	act.sa_flags = SA_ONSTACK | SA_RESETHAND;

	sigaltstack(&st, NULL);
	sigaction(SIGSEGV, &act, NULL);
	sigaction(SIGBUS, &act, NULL);
	sigaction(SIGABRT, &act, NULL);

	// fork if needed and drop privileges as requested.
	// this must be done before we do anything with pthreads
	if (daemonize && unix_daemonize(m_lpThreadMonitor->lpConfig, m_lpThreadMonitor->lpLogger))
		goto exit;
	if (!daemonize)
		setsid();
	if (unix_create_pidfile(argv[0], m_lpThreadMonitor->lpConfig, m_lpThreadMonitor->lpLogger, false) < 0)
		goto exit;
	if (unix_runas(m_lpThreadMonitor->lpConfig, m_lpThreadMonitor->lpLogger))
		goto exit;

	// Init exit threads
	pthread_mutex_init(&m_hExitMutex, NULL);
	pthread_cond_init(&m_hExitSignal, NULL);

		hr = running_service(szPath);


exit:
	if(m_lpThreadMonitor)
		deleteThreadMonitor(m_lpThreadMonitor, true);

	return hr == hrSuccess ? 0 : 1;
}

HRESULT running_service( char* szPath)
{
	HRESULT			hr = hrSuccess;
	ECScheduler*	lpECScheduler = NULL;
	unsigned int	ulInterval = 0;
	bool			bMapiInit = false;
	
	hr = MAPIInitialize(NULL);
	if (hr != hrSuccess) {
		m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Unable to initialize MAPI");
		goto exit;
	}
	bMapiInit = true;

	lpECScheduler = new ECScheduler(m_lpThreadMonitor->lpLogger);

	ulInterval = atoi(m_lpThreadMonitor->lpConfig->GetSetting("quota_check_interval", NULL, "15"));
	if (ulInterval == 0)
		ulInterval = 15;

	m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Starting zarafa-monitor version " PROJECT_VERSION_MONITOR_STR " (" PROJECT_SVN_REV_STR "), pid %d", getpid());

	// Add Quota monitor
	hr = lpECScheduler->AddSchedule(SCHEDULE_MINUTES, ulInterval, ECQuotaMonitor::Create, m_lpThreadMonitor);
	if(hr != hrSuccess) {
		m_lpThreadMonitor->lpLogger->Log(EC_LOGLEVEL_FATAL, "Unable to add quota monitor schedule");
		goto exit;
	}

	pthread_mutex_lock(&m_hExitMutex);

	pthread_cond_wait(&m_hExitSignal, &m_hExitMutex);

	pthread_mutex_unlock(&m_hExitMutex);

exit:
	if (lpECScheduler)
		delete lpECScheduler;
		
	if (bMapiInit)
		MAPIUninitialize();
		
	return hr;
}
